compiler/circle-mpqsolver/src/bisection/BisectionSolver.cpp

```diff
@@ -0,0 +1,211 @@
+#include "DepthParameterizer.h"
+#include "ErrorMetric.h"
+
+#include <luci/ImporterEx.h>
+#include <luci/Log.h>
+
+#include <cmath>
+#include <iostream>
+
+using namespace mpqsolver::bisection;
+
+namespace
+{
+
+std::unique_ptr<luci::Module> read_module(const std::string &path)
+{
+  luci::ImporterEx importerex;
+  auto module = importerex.importVerifyModule(path);
+  if (module.get() == nullptr)
+  {
+    std::cerr << "Failed to load " << path << std::endl;
+    return nullptr;
+  }
+
+  return module;
+}
+
+} // namespace
+
+BisectionSolver::BisectionSolver(const std::string &input_data_path, float qerror_ratio,
+                                 const std::string &input_quantization,
+                                 const std::string &output_quantization)
+  : MPQSolver(input_data_path, qerror_ratio, input_quantization, output_quantization)
+{
+  _quantizer = std::make_unique<Quantizer>(_input_quantization, _output_quantization);
+}
+
+float BisectionSolver::evaluate(const DatasetEvaluator &evaluator, const std::string &flt_path,
+                                const std::string &def_quant, LayerParams &layers)
+{
+  auto model = read_module(flt_path);
+  // get fake quantized model for evaluation
+  if (!_quantizer->fake_quantize(model.get(), def_quant, layers))
+  {
+    throw std::runtime_error("Failed to produce fake-quantized model.");
+  }
+
+  return evaluator.evaluate(model.get());
+}
+
+void BisectionSolver::algorithm(Algorithm algorithm) { _algorithm = algorithm; }
+
+std::unique_ptr<luci::Module> BisectionSolver::run(const std::string &module_path)
+{
+  LOGGER(l);
+
+  auto module = read_module(module_path);
+
+  float min_depth = 0.f;
+  float max_depth = 0.f;
+  NodeDepthType nodes_depth;
+  if (compute_depth(module.get(), nodes_depth, min_depth, max_depth) !=
+      ParameterizerResult::SUCCESS)
+  {
+    std::cerr << "Invalid graph for bisectioning" << std::endl;
+    return nullptr;
+  }
+
+  std::unique_ptr<MAEMetric> metric = std::make_unique<MAEMetric>();
+  DatasetEvaluator evaluator(module.get(), _input_data_path, *metric.get());
+
+  LayerParams layer_params;
+  float int16_qerror =
+    evaluate(evaluator, module_path, "int16" /* default quant_dtype */, layer_params);
+  VERBOSE(l, 0) << "Full int16 model quantization error " << int16_qerror << std::endl;
+
+  float uint8_qerror =
+    evaluate(evaluator, module_path, "uint8" /* default quant_dtype */, layer_params);
+  VERBOSE(l, 0) << "Full uint8 model quantization error " << uint8_qerror << std::endl;
+
+  if (int16_qerror > uint8_qerror)
+  {
+    throw std::runtime_error("Q8 model's qerror is less than Q16 model's qerror.");
+  }
+
+  _qerror = int16_qerror + _qerror_ratio * std::fabs(uint8_qerror - int16_qerror);
+  VERBOSE(l, 0) << "Target quantization error " << _qerror << std::endl;
+
+  if (int16_qerror > _qerror)
+  {
+    std::cerr << "ERROR: Qerror of full Q16 model is higher than the target qerror." << std::endl;
+    return nullptr;
+  }
```

Comment:
Hmm.. this check seems redundant.
```suggestion
```

Reply:
This was made to provide some guard in case `qerror_ratio < 0` or user can submit `target_qerror` directly. Currently it is redundant.